File: learner1.hs
{-
  Programmer: Matthew Healy <mhrh3@mst.edu>
      Course: CS5001 Deep Learning
  Assignment: HW #1, Linear Regression Learner
   Professor: Dr. Ricardo Morales
-}
import System.Random
import System.IO
import Data.List
import Data.Char

sumSquareError :: (Num a) => a -> a -> [(a, a)] -> a
sumSquareError w0 w1 [es] =
  sum [ ( ( snd e ) - ( yHat w0 w1 e ) )^2 | e <- es ]

yHat :: (Num a) => [a] -> [a] -> a
yHat ws xs = let xs' =
  sum $ [(ws !! i) * (xs !! i) | i <- [1..(len xs)]]

linError :: (Num a) => [a] -> ([a] -> a) -> a
linError ws (xs y) = y - (yHat ws xs)

learn :: (Num a) => [a] -> [a] -> a -> [a]
learn weights examples eta =
  learn' weights examples eta 5000
  where learn' ws es eta count
          | count == 0 = ws
          | otherwise = let s = len xs in
            [(ws !! i) + eta * (linError ws es !! i) * (es !! i) | i <- [1..s]]

getExamples :: String -> [([Num], Num)]
getExamples s =
  getExamples' (splitOn "\n" s) []
  where getExamples' s es
          | s == [] = es
          | otherwise =
            let
              (xString, yString) 
            in
              [(xs, y)] ++ getExamples' (tail s) es



main = do
  contents <- readFile "trash.txt"
  es = getExamples contents
  seed = --wat now???
  eta = 2 -- arbitrarily pick learning rate
  g <- getStdGen
  ws = take 2 $ randomRs (0, 100) g -- initialize weights
  ws' = learn ws es eta
  sse = sumSquareError ws' es
  hdr = ("CS-5001: HW#1 \nProgrammer: Matthew Healy <mhrh3>\n\n"
  ++ "TRAINING\n"
  ++ "Using random seed = " ++ (show seed) ++ "\n"
  ++ "Using learning rate eta = " ++ (show eta) ++ "\n"
  ++ "After 5000 iterations:\n"
  ++ "Weights:\n"
  ++ "w0 = " ++ (show ws' !! 0) ++ "\n"
  ++ "w1 = " ++ (show ws' !! 1) ++ "\n"
  ++ "\n"
  ++ "VALIDATION\n"
  ++ "Sum-of-Squares Error = " ++ (show sse))
  putStrLn hdr

File: learner1.py
#! /usr/bin/python3
import sys
from random import random, randint, seed

ITERATIONS = 5000
ETA = 0.0005


class Example:
    def __init__(self, xs, y):
        self.x = xs + [1]
        self.y = y


def sum_of_squares(E, w):
    sum = 0
    for e in E:
        sum += (e.y - y_hat(w, e)) ** 2
    return sum


def y_hat(w, e):
    yh = 0
    for i in range(len(e.x)):
        yh += w[i] * e.x[i]
    return yh


def getExamples(input_file):
    examples = []
    with open(input_file, "r") as f:
        ns = f.readline().split()
        ns = [float(n) for n in ns]
        y = ns[-1]
        xs = ns[:-1]
        examples.append(Example(xs, y))
    return examples


def learn(w, E, eta):
    for i in range(ITERATIONS):
        for k in range(len(E)):
            yh = y_hat(w, E[k])
            delta = E[k].y - yh
            for j in range(len(w)):
                w[j] += eta * delta * E[k].x[j]
    return w


def main():
    if len(sys.argv) > 1:
        my_seed = int(sys.argv[1])
    else:
        my_seed = randint(0, 1000000)
    seed(my_seed)
    w = [randint(0, 100), randint(0, 100)]
    eta = ETA
    E = getExamples("trashdata.txt")
    Ev = getExamples("moretrashdata.txt")
    w = learn(w, E, eta)
    output = """\
CS-5001: HW#1
Programmer: Matthew Healy <mhrh3@mst.edu>

TRAINING
Using random seed = {}
Using learning rate eta = {}
After {} iterations:
Weights:
w0 = {}
w1 = {}

VALIDATION
Sum-of-Squares Error = {}""".format(
        my_seed, eta, ITERATIONS, w[0], w[1], sum_of_squares(Ev, w)
    )
    print(output)


if __name__ == "__main__":
    main()
